annotated.go

```diff
@@ -144,6 +144,82 @@ type paramTagsAnnotation struct {
 }
 
 var _ Annotation = paramTagsAnnotation{}
+var (
+	errTagSyntaxSpace            = errors.New(` multiple tags are not separated by space`)
+	errTagKeySyntax              = errors.New(" tag key is invalid, Use group, name or optional as tag keys")
+	errTagValueSyntaxQuote       = errors.New(`tag value should start with double quote. i.e. key:"value" `)
+	errTagValueSyntaxEndingQuote = errors.New(`tag value should end in double quote. i.e. key:"value" `)
+)
+
+// Given a tag with key, value pair, they must be separaed by a space.
+// Eg: `group:"some" optional:"true"`
+// return an error if in the case of multiple tags, they aren't separated
+// space. Return nil on success.
+func verifyTagsSpaceSeparated(numTags int, tag string) error {
+	if numTags > 0 && tag != "" && tag[0] != ' ' {
+		return errTagSyntaxSpace
+	}
+	return nil
+}
+
+// Given a value from a tag verify that it is contained between double quotes
+// Return an error if the invalid quote format is used for tag value.
+// On sucess return the remaining tag with nil error
+func verifyValueQuote(value string) (string, error) {
+	// starting quote should be a double quote
+	if value[0] != '"' {
+		return "", errTagValueSyntaxQuote
+	}
+	// validate tag value is within quotes
+	i := 1
+	for i < len(value) && value[i] != '"' {
+		if value[i] == '\\' {
+			i++
+		}
+		i++
+	}
+	if i >= len(value) {
+		return "", errTagValueSyntaxEndingQuote
+	}
+	return value[i+1:], nil
+
+}
+
+// Given a string tag the function checks if the tag follows valid struct
+// format and returns an error if it's invalid. (i.e. not following
+// tag:"value" space-separated list )
+// Currently dig accepts only 'name', 'group', 'optional' as valid tag keys
+func verifyAnnotateTag(tag string) error {
+	numTags := 0
+	for ; tag != ""; numTags++ {
+		validKeys := map[string]bool{"group": true, "optional": true, "name": true}
+		if err := verifyTagsSpaceSeparated(numTags, tag); err != nil {
+			return err
+		}
+		i := 0
+		// return if tag is empty
+		if strings.TrimSpace(tag) == "" {
+			return nil
+		}
+
+		// parsing the key i.e. till reaching colon :
+		for i < len(tag) && tag[i] != ':' {
+			i++
+		}
+		key := strings.TrimSpace(tag[:i])
+		// if key is not equal to group name or optional
+		if _, ok := validKeys[key]; !ok {
+			return errTagKeySyntax
+		}
+		value, err := verifyValueQuote(tag[i+1:])
```

Comment:
This code has an interesting struct where the value passed through a set of filters and runs through the same set with the remaining processed values. Have you experimented to see if the code is simplified if implemented recursively?